src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,147 @@
+            final Node<T> nextNode = getNodeByIndex(index);
+            final Node<T> prevNode = nextNode.prev;
+            final Node<T> newNode = new Node<>(prevNode, value, nextNode);
+            nextNode.prev = newNode;
+            if (prevNode == null) {
+                head = newNode;
+            } else {
+                prevNode.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element: list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = getNodeByIndex(index);
+        T currentItem = currentNode.item;
+        currentNode.item = value;
+        return currentItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        return unlink(getNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if (object == null && currentNode.item == null
+                    || object != null && object.equals(currentNode.item)) {
+                remove(i);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public Node<T> getNodeByIndex(int index) {
+        Node<T> currentNode;
+        if (index < (size >> 1)) {
+            currentNode = head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            currentNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                currentNode = currentNode.prev;
+            }
+        }
+        return currentNode;
+    }
+
+    private T unlink(Node<T> node) {
+        final T element = node.item;
+        final Node<T> next = node.next;
+        final Node<T> prev = node.prev;
+        if (prev == null) {
+            head = next;
+        } else {
+            prev.next = next;
+            node.prev = null;
+        }
+
+        if (next == null) {
+            tail = prev;
+        } else {
+            next.prev = prev;
+            node.next = null;
+        }
+        size--;
+        return element;
+    }
+
+    private void checkIndex(int index) {
+        if (!(index >= 0 && index < size)) {
+            throw new ArrayIndexOutOfBoundsException("Index: " + index
+                    + " , out of bound, array length - " + size);
+        }
+    }
```

Comment:
```suggestion
    private void checkIndex(int index) {
        if (index < 0 || index >= size) {
```
such a record is easier to read